### PR TITLE
Half Normal Distribution (and inverse error function)

### PR DESCRIPTION
**Pull request by cshenton**
## Inverse Error Function

Added `erfinv` as a simple function of `ndtri` (the api mirrors the approach taken in `scipy.special`. Mainly to prevent distributions with methods in terms of the inverse error function reimplementing their own wrappers around `ndtri`.

## Half Normal Distribution
```python
X ~ Normal(0.0, scale)
Y = |X|
# then Y ~ HalfNormal(scale)
```

Added the Half Normal distribution to `contrib.distributions`. Main things to look at is how I'm dealing with the pdf discontinuity.

A discontinuity in the pdf that sends values to 0.0 is easy to represent in `prob`, (by multiplying my a tensor with 0.0s at the relevant positions) than `log_prob` (which requires masking -inf wherever x < 0). I'd be interested if anyone has any suggestions for correct ways to implement the latter, as I think this is a case where `log_prob` is more numerically stable (along the support of the distribution that is).

### CODE REVIEW

**Comment by cshenton** [comment]
Sanity checks failure seems to be unrelated to the content of the PR:
```
The command '/bin/sh -c /install/install_deb_packages.sh' returned a non-zero code: 100
ERROR: docker build failed. Dockerfile is at /var/lib/jenkins/workspace/tensorflow-pull-requests-sanity/tensorflow/tools/ci_build/Dockerfile.cpu
```

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by ebrevdo** [comment]
The details mostly look good.  Is there a reason you didn't use the [absolutevalue Bijector](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distributions/python/ops/bijectors/absolute_value_impl.py#L34) to implement this class?  It would take care of sampling and log_prob for you.

**Comment by cshenton** [comment]
Thank for bringing that to my attention. If that's available, I'll implement a the `FoldedNormal` with `HalfNormal` as a special case.

As an aside, the docs state that bijectors implement bijective differentiable transforms, but the absolute value function is not a bijection. Am I right in guessing bijectors is moving towards being a more general set of transforms for distributions?

**Comment by ebrevdo** [comment]
Yes; it's an extension now that bijector inverses (and inverse log det
jacobian functions) can return tuples - under the assumption that the
preimage is always len(tuple)-valued.  I believe that the special case of
t=0 should continue to work for HalfNormal due to the way
TransformedDistribution treats these values, but if not then you may have
to special case that value.

**Comment by cshenton** [comment]
Fixed things up. Tests were previously failing since the scipy `halfnorm`'s first arg was not `scale` as I assumed, but `loc`, where loc is both the mean of the base distribution and the fold point.

Also I didn't end up using the bijector and subclassing `TransformedDistribution`, because the static shape of the outputs of `prob` and `log_prob` were both `<unknown>`.

**Comment by cshenton** [comment]
Could someone manually trigger the test run? Container build failed on the sanity tests for some reason.
```
ERROR: docker build failed. Dockerfile is at /var/lib/jenkins/workspace/tensorflow-pull-requests-sanity/tensorflow/tools/ci_build/Dockerfile.cpu
Build was aborted
Aborted by unknown
Unable to get pull request builder trigger!!
```

**Comment by ebrevdo** [comment]
@tensorflow-jenkins test this please.

**Comment by cshenton** [comment]
Looks like one of the macOS tests timed out, so tests will need to be rerun. Apologies for the trouble. 
```
//tensorflow/python/kernel_tests:slice_op_test                          TIMEOUT in 301.0s
```

**Review by ebrevdo — APPROVED** [review]

File: tensorflow/contrib/distributions/python/kernel_tests/half_normal_test.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+            variables.global_variables_initializer().run()
+            self.assertAllFinite(value)
+            self.assertAllFinite(grads[0])
+
+  def testHalfNormalEntropy(self):
+    with self.test_session():
+      scale = np.array([[1.0, 2.0, 3.0]])
+      halfnorm = hn_lib.HalfNormal(scale=scale)
+
+      expected_entropy = 0.5 * np.log(np.pi * scale ** 2.0 / 2.0) + 0.5
```
Comment (ebrevdo) [inline tensorflow/contrib/distributions/python/kernel_tests/half_normal_test.py]: add a reference for this calculation?

Comment (cshenton) [reply tensorflow/contrib/distributions/python/kernel_tests/half_normal_test.py]: Will do. Is a link to wikipedia fine? There's the 1961 paper on the folded normal distributions but that's paywalled so less easy to check.

**Comment by ebrevdo** [comment]
Wiki is fine

**Comment by cshenton** [comment]
Do the remaining checks need to be triggered, or is the Ubuntu CC just taking a really long time?

**Pull request merged**